### PR TITLE
Case class copy and apply inherit access modifiers from constructor (under -Xsource:3)

### PR DESCRIPTION
**Pull request by Jasper-M**
Backport from dotty:

- If a case class constructor is `private` or `private[foo]`: the synthesized `copy` and `apply` methods will have the same access modifier.
- If a case class constructor is `protected` or `protected[foo]`: the synthesized `copy` method will have the same access modifier. The synthesized `apply` method will remain `public`, because `protected` doesn't make sense in an `object`.

Obviously, if a user defines a custom `copy` or `apply` method, that one—including its access modifier—will take precedence.

See the accompanying [pos](https://github.com/scala/scala/blob/d6774a8bc7a0ee02b79a946f85f6dd344eb45b5f/test/files/pos/caseclass_private_constructor.scala) and [neg](https://github.com/scala/scala/blob/d6774a8bc7a0ee02b79a946f85f6dd344eb45b5f/test/files/neg/caseclass_private_constructor.scala) tests.

Only enabled under ~~`-Xsource:2.14` and by extension under~~ `-Xsource:3`.

### CODE REVIEW

**Comment by Jasper-M** [comment]
The 3 failing tests all look expected to me.

**Comment by Jasper-M** [comment]
Added some changes to bring the PR in line with lampepfl/dotty#5830 and fixes for corresponding dotty issues lampepfl/dotty#5857 and lampepfl/dotty#5859.

**Comment by Jasper-M** [comment]
@adriaanm All those private constructors in `pos/userdefined_apply.scala`, are they there to avoid that calls to `apply` get transformed to calls to the constructor? In which case it would be fine to make them protected instead?

**Comment by adriaanm** [comment]
> @adriaanm All those private constructors in `pos/userdefined_apply.scala`, are they there to avoid that calls to `apply` get transformed to calls to the constructor? In which case it would be fine to make them protected instead?

IIRC it's to control who can create an instance of a class -- if this is what you want,  a private constructor is useless if your apply is public, so basically the same as what your PR offers by default :-)

**Comment by Jasper-M** [comment]
@adriaanm Err... Not following 100% what you're saying. Some of those tests seem to rely specifically on the generated `apply` *not* being private.

But anyway I think this PR is ready for review and discussion on how to proceed (SIP or no SIP, -Xsource flag or no flag, ...).

Status of this PR right now: 
- private and protected, both qualified and unqualified, get inherited by the copy method from the constructor
- (un)qualified private gets inherited by the apply method from the constructor
- (un)qualified protected constructor results in normal public apply method, as seen in dotty
- case classes with a (un)qualified private constructor will have a companion object that does not extend a FunctionN trait.

I reviewed all failing tests and the bugs they are defending against. In all cases the right thing to do seemed to me turning private constructors into protected. That way `apply` is still public _and_ calls to `apply` will not get rewritten into calls to the constructor.

**Comment by eed3si9n** [comment]
> But anyway I think this PR is ready for review and discussion on how to proceed (SIP or no SIP, -Xsource flag or no flag, ...).

@smarter Is there a SIP for this on Dotty's change?

File: test/files/pos/caseclass_private_constructor.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
+case class A private (i: Int)
+object A {
+  def a = A(1).copy(2) // apply and copy are accessible in companion
+}
```
Comment (eed3si9n) [inline test/files/pos/caseclass_private_constructor.scala:6]: Maybe there should be a test for variant of `A` that checks that I can override `apply` in companion to make it public?

**Comment by smarter** [comment]
> @smarter Is there a SIP for this on Dotty's change?

There isn't.

**Comment by tpolecat** [comment]
I wasn't aware of this PR until just now. It would be really nice to get this into Scala 2 so we can get away from `sealed abstract case class` (although I admit it's probably my favorite trick).

**Comment by Jasper-M** [comment]
Added some tests based on @eed3si9n's comment. And rebased in the hopes of fixing MIMA failures in travis.

**Comment by Jasper-M** [comment]
Ok, I get it now. MIMA complains because some apply and copy methods became private.

**Comment by Jasper-M** [comment]
I put the change behind the `-Xsource:2.14` flag, since this can't go in 2.13 anyway. That should also fix the MIMA issues.

**Review by lrytz — APPROVED** [review]
This looks really good! Implementation, testing, all 👍.

I think this doesn't need a SIP.

My only concern is that it's a breaking change, but I cannot think of anything that we can do in 2.13 to aid / anticipate migration.

**Comment by Jasper-M** [comment]
Can this be considered for 2.13.2?

**Comment by lrytz** [comment]
Under `-Xsource:2.14` for sure. I didn't merge it so @adriaanm can take another look.

**Comment by adriaanm** [comment]
Sorry for the delay. I will do my best to look soon. In principle it all looks good.

**Comment by adriaanm** [comment]
Sorry for lagging on this. I've asked Jason to take a look instead. He's been reworking this part of namers.

**Comment by SethTisue** [comment]
@retronym do you intend to try to get this over the finish line for 2.13.2?

**Comment by lrytz** [comment]
Let's merge this, it's looking good

**Pull request merged**